### ISSUE TITLE
Concurrently Local-Only Wifi Support

### ISSUE THREAD

**Issue by bencorrado** [home-assistant/android#3961]
**Is your feature request related to a problem? Please describe.**

Home Assistant is designed with a focus on local control.  If a Home Assistant server is on a local only network (no internet access) in recent versions of Android (tested in 13 & 14, but I believe 12 is also an issue), the Android device will drop the WiFi network and only stay connected to the mobile network if one is present.  This forces a user to disable the LTE on their phone to connect to the Wi-Fi network and control their Home Assistant Server.  This is a very frustrating user experience for offline servers of Home Assistant when using the Android app with LTE connection.

**Describe the solution you'd like**

I would like to take advantage of [Concurrent local-only and internet connection](https://source.android.com/docs/core/connect/wifi-sta-sta-concurrency#local-only) in more recent Android SDKs.  This allows for the internet to be provided by LTE or a primary WiFi network, while the app connects to a secondary WiFi that does not have internet access for local only access.  It was designed for IoT support, but should work well with an offline Home Assistant server.

We would allow a checkbox next to Home network WiFi SSD entry screen to allow users to designate if a Wi-Fi network may only have local access (no internet).  

We could then use the SSID from the local access only network to connect to the server by using WifiNetworkSpecifier.Builder() similar to this:

```
fun connectToLocalOnlyNetwork(ssid: String, context: Context) {
    val wifiNetworkSpecifier = WifiNetworkSpecifier.Builder()
        .setSsid(ssid)
        .build()

    val networkRequest = NetworkRequest.Builder()
        .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
        .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN) // Ensure it's not a VPN
        .setNetworkSpecifier(wifiNetworkSpecifier)
        .build()

    val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            super.onAvailable(network)
            // Local-only WiFi is connected
        }

        override fun onUnavailable() {
            super.onUnavailable()
            // Failed to connect
        }
    }

    connectivityManager.requestNetwork(networkRequest, networkCallback)
}

```

We would want to enable the checkbox to support this only on devices with Android 10 or later and maybe only for devices where WifiManager#isStaConcurrencyForLocalOnlyConnectionsSupported() is true.  However, that might just need to be a warning that selecting local support will disconnect their other WiFi in order to connect to Home Assistant if it is false.

I think the primary issue to determine is how to create the connection to the local only WiFi network for just regular metrics updates and not drain the battery with a WiFi scan if it is disconnected.  Testing is certainly required for that.

**Describe alternatives you've considered, if any**

The primary alternative here is turning off LTE access to connect to a WiFi network.  We could also consider Concurrent restricted and internet connection or a BindNetwork, but I think both of these have drawbacks that Concurrent local-only and internet connection do not have.

**Additional context**

This is related to #675 which was closed due to a lack of support at the time from Android SDK to support this.

**Comment by nima-1102** [comment]
+1

**Comment by kquinsland** [comment]
+1 for this as well. 

My ISP had a prolonged outage which meant DNS stopped working and thus Android phone quickly shoveled all data over LTE.  I had to explicitly disable mobile data before I could load home assistant web interface.

**Comment by bluesk1d** [comment]
Yes, please. Internet is down currently and have to switch mobile data on and off to get to HA/the Internet.

**Comment by hedgepigdaniel** [comment]
> the Android device will drop the WiFi network and only stay connected to the mobile network if one is present

This is not the case for me. In a scenario with no-internet on the wifi and a mobile network with internet, wifi remains connected and shows "Connected / No internet access" in the wifi settings menu.

But still, as detailed in https://github.com/home-assistant/android/issues/675, addresses on the local wifi network are not reachable in this scenario (HA instance is not reachable from Chrome, Firefox or the HA app).

So I think the first problem to solve is how to get the home assistant app to successfully connect to the server if it is on a local wifi network and the mobile network is used for internet. Maybe this is possible without the app knowing/caring about the specific wifi network? It seems that this problem needs to be solved, regardless of whether the wifi network connection is handled by the app or by the user.

> This is a very frustrating user experience for offline servers of Home Assistant when using the Android app with LTE connection.

100%. This is a pretty big problem IMO for a home automation software that is explicitly designed to work locally.

**Comment by hedgepigdaniel** [comment]
See also https://android.stackexchange.com/questions/258380/cannot-access-to-local-devices-when-the-connection-to-the-internet-is-interrupte?newreg=57c587b0af4a468a8f63b7f9c876ef0d

**Comment by zcattacz** [comment]
https://www.jrklein.com/2019/04/06/using-mobile-internet-data-while-connected-to-a-wifi-network-ios-android/

Is it possible to configure a wifi network to use static IP setting and clear its default gateway ? I can't do it on my Android 9 phone's system WiFi Setting UI. It would't like me save. Not sure if this is possible on other version or through a programmatic interface ?

**Comment by zcattacz** [comment]
issues with wifinetworkspecifier API in 2022  https://stackoverflow.com/questions/60660236/android-q-wifi-connection-via-wifinetworkspecifier-lose-connection-immediately-a

**Comment by bencorrado** [comment]
I have opened an Upwork project to try and get this fixed.  If anyone want to help sponsor it let me know

https://www.upwork.com/jobs/~021886560595067584015

**Comment by bencorrado** [comment]
@jpelgrom If you are interested in tackling this, I am willing to compensate!
